geps/gep-1651.md

```diff
@@ -0,0 +1,154 @@
+# GEP-1651: GatewayAddress Scope
+
+* Issue: [#1651](https://github.com/kubernetes-sigs/gateway-api/issues/1651)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Allow users to configure a GatewayAddress so that it is only accessible within
+a specific scope (ie. public/cluster local)
+
+## Goals
+
+- Cluster local gateways can be created in a declarative way `ClusterLocal` and
+  are only accessible within a single cluster
+- Define a `Public` scope for addresses that are accessible to the internet
+- Implementations can define & support their own scopes
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an 
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal" 
+(service.namespace.svc) and "external" (service.namespace.example.com). 
+Gateways using the same GatewayClass on the cluster, but ensure that the 
+"internal" service is only reachable within the cluster. This would greatly 
+simplify deployment for users over the instructions we have today.
+
```

Comment:
IMO this use case can be solved through GatewayClass parametersRef. For example:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: GatewayClass
metadata:
  name: example-class
spec:
  controllerName: my.example.io/gatewayclass-controller
  parametersRef:
    name: example-config
    group: config.example.io
    kind: GatewayClassParams
---
apiVersion: config.example.io/v1alpha1
kind: GatewayClassParams
metadata:
  name: example-config
spec:
  networkPublishing:
    types:
    - KubeClusterIP
    - KubeLoadBalancer
```

The controller would create 2 Services (LoadBalancer and ClusterIP) when a Gateway of this GatewayClass is created. If the Gateway specifies `gateway.spec.addresses` that are not `IPAddressType`, or addresses are not available, the Gateway should report the "Invalid=True" status condition with an appropriate reason.

__Note:__ The example `networkPublishing` API above can be a union type to expose type-specific configuration.

Reply:
For the record, I agree with @danehans here, I think this option deserves significantly more attention in this GEP.